feat(types): add `InferAttributes` utility type

ephys opened the pull request:
### Pull Request Checklist

- [x] Have you added new tests to prevent regressions?
- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you update the typescript typings accordingly (if applicable)?
- [ ] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)?

### Description Of Change

This PR introduces a few new utility types:
- `InferAttributes<M, opts>`: Used to drastically reduce the boilerplate needed to define Model Attributes in the class-based approach.
- `InferCreationAttributes<M, opts>`: Like `InferAttributes` but fields tagged with `CreationOptional<x>` are optional.
- `CreationOptional<x>`: A branded type used to mark an attribute as optional in Creation Attributes.
- `NonAttribute<x>`: A branded type used to omit a property from the attribute list.
- `Attributes<M>`: An alias for `M['_attributes']` for public use. Also forward compatible with planned v7 changes.
- `CreationAttributes<M>`: Same as `Attributes` but types that accept `undefined` are also `optional`.

To reduce risk of collision between `Model` and its subclasses, `Model#_attributes` and `Model#_creationAttributes` have been deprecated (I'd like to replace them with symbols in v7). Use `Attributes<Model>` and `CreationAttributes<Model>` instead.

In v7, we should replace these two with non-exported unique symbols.

WikiRik commented on docs/manual/other-topics/typescript.md:
```diff
@@ -296,6 +296,225 @@ async function doStuffWithUserModel() {
+
+// 'projects' is excluded as it's not an attribute, it's an association.
+class User extends Model<AttributesOf<User, 'projects'>> {
+  // can be undefined during creation when using `autoIncrement`
```
```suggestion
  // id can be undefined during creation when using `autoIncrement`
```

To make sure people know where it refers to

WikiRik commented on docs/manual/other-topics/typescript.md:
```diff
@@ -296,6 +296,225 @@ async function doStuffWithUserModel() {
+  declare preferredName: string | null; // for nullable fields
+
+  // timestamps!
+  // can be undefined during creation
```
```suggestion
  // createdAt can be undefined during creation
```

WikiRik commented on docs/manual/other-topics/typescript.md:
```diff
@@ -296,6 +296,225 @@ async function doStuffWithUserModel() {
+  declare ownerId: number;
+  declare name: string;
+
+  // can be undefined during creation
```
```suggestion
  // createdAt can be undefined during creation
```

WikiRik commented on docs/manual/other-topics/typescript.md:
```diff
@@ -296,6 +296,225 @@ async function doStuffWithUserModel() {
+
+  // can be undefined during creation
+  declare readonly createdAt?: Date;
+  // can be undefined during creation
```
```suggestion
  // updatedAt can be undefined during creation
```

WikiRik commented on types/test/typescriptDocs/ModelInitWithAttributesOf.ts:
```diff
@@ -0,0 +1,197 @@
+/**
+ * Keep this file in sync with the code in the "Using with `AttributesOf`" section in typescript.md
```
This note is only for the contributors of Sequelize right? Even though, we should put the complete path to the file

ephys replied:
Done, as well as the other files like this

WikiRik commented on types/test/typescriptDocs/ModelInitWithAttributesOf.ts:
```diff
@@ -0,0 +1,197 @@
+
+// 'projects' is excluded as it's not an attribute, it's an association.
+class User extends Model<AttributesOf<User, 'projects'>> {
+  // can be undefined during creation when using `autoIncrement`
```
```suggestion
  // id can be undefined during creation when using `autoIncrement`
```

WikiRik commented on types/test/typescriptDocs/ModelInitWithAttributesOf.ts:
```diff
@@ -0,0 +1,197 @@
+  declare preferredName: string | null; // for nullable fields
+
+  // timestamps!
+  // can be undefined during creation
```
```suggestion
  // createdAt can be undefined during creation
```

WikiRik commented on types/test/typescriptDocs/ModelInitWithAttributesOf.ts:
```diff
@@ -0,0 +1,197 @@
+  // timestamps!
+  // can be undefined during creation
+  declare readonly createdAt?: Date;
+  // can be undefined during creation
```
```suggestion
  // updatedAt can be undefined during creation
```

WikiRik commented on types/test/typescriptDocs/ModelInitWithAttributesOf.ts:
```diff
@@ -0,0 +1,197 @@
+  declare ownerId: number;
+  declare name: string;
+
+  // can be undefined during creation
```
```suggestion
  // createdAt can be undefined during creation
```

WikiRik commented on types/test/typescriptDocs/ModelInitWithAttributesOf.ts:
```diff
@@ -0,0 +1,197 @@
+
+  // can be undefined during creation
+  declare readonly createdAt?: Date;
+  // can be undefined during creation
```
```suggestion
  // updatedAt can be undefined during creation
```

WikiRik commented on types/test/typescriptDocs/ModelInitWithAttributesOf.ts:
```diff
@@ -0,0 +1,197 @@
+  declare userId: number;
+  declare address: string;
+
+  // can be undefined during creation
```
```suggestion
  // createdAt can be undefined during creation
```

WikiRik commented on types/test/typescriptDocs/ModelInitWithAttributesOf.ts:
```diff
@@ -0,0 +1,197 @@
+
+  // can be undefined during creation
+  declare readonly createdAt?: Date;
+  // can be undefined during creation
```
```suggestion
  // updatedAt can be undefined during creation
```

ephys commented:
No no, thank you for the review :) I did parts of it in a hurry to go eat and didn't notice most of the things you mentioned

ephys commented on docs/manual/other-topics/typescript.md:
```diff
@@ -296,6 +296,225 @@ async function doStuffWithUserModel() {
+  If one of your attributes shares the same name as one of the properties of `Model`, change its name.
+  Doing this is likely to cause issues anyway.
+
+**NOTE:** Keep the following code in sync with `/types/test/typescriptDocs/ModelInitWithAttributesOf.ts` to ensure it typechecks correctly.
```
It's meant for us yes. I've hidden it in a HTML comment so it's only visible in the raw md file. I've changed existing notes like this too.

ephys commented:
The mapped types I'm using in this PR require TS >= 4.1. We'd need to drop support for TS <= 4.1 to merge this.

I'm on board with it. It's more than a year old now. The question is: do we want to consider that a semver major or minor?

WikiRik commented:
> The mapped types I'm using in this PR require TS >= 4.1. We'd need to drop support for TS <= 4.1 to merge this.
> 
> I'm on board with it. It's more than a year old now. The question is: do we want to consider that a semver major or minor?

We had this discussion in another PR as well and I had difficulty with finding a good support plan for TypeScript versions. One that I did find was [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped#support-window) which tests on TypeScript versions that are less than 2 years old. But if you find other examples I would be fine with shortening the support window.

Since TypeScript releases come out approximately every three months and the minority of users use TypeScript, I would make this a semver minor change.

ephys commented:
We can use `typesVersions` to expose different typing files based on TS version. I can try to do something with it.
Alternatively we drop support for versions `< 4.1` 8 months early (`4.0` having been released in august 2020)

ephys commented:
Before we merge this, I'm also thinkings of ways to solve the `TCreationAttributes` part:

I have a solution that uses a type as branding to mark attributes as optional in TCreationAttributes:

<img width="873" alt="image" src="https://user-images.githubusercontent.com/1280915/148244120-133f98c2-5283-44de-b2d6-a519513fc908.png">

It's a bit hacky but so far it works.

If we implement this as well as https://github.com/sequelize/meetings/issues/14, in v7 we should be able to simplify the Model definition to simply:

```
class User extends Model<User> {}
```

`Model` should be able to extract all the necessary information from the `User` class directly

AllAwesome497 commented on docs/manual/other-topics/typescript.md:
```diff
@@ -296,12 +293,228 @@ async function doStuffWithUserModel() {
+// 'projects' is excluded as it's not an attribute, it's an association.
+class User extends Model<AttributesOf<User, 'projects'>> {
+  // id can be undefined during creation when using `autoIncrement`
+  declare id?: number;
```
Is this `declare` needed?

ephys replied:
Currently yes, due to https://sequelize.org/v6/manual/model-basics.html#caveat-with-public-class-fields

Also relevant is this https://github.com/sequelize/meetings/issues/16

AllAwesome497 reviewed (commented):
Not an issue with this PR, but I think these associations are a bit too complex - its very easy to forget some.

ephys commented on docs/manual/other-topics/typescript.md:
```diff
@@ -296,12 +293,228 @@ async function doStuffWithUserModel() {
+  // timestamps!
+  // createdAt can be undefined during creation
+  declare readonly createdAt?: Date;
+  // updatedAt can be undefined during creation
+  declare readonly updatedAt?: Date;
```
It's a copy of the existing documentation but I agree we could simplify this to only the important bits. It's up to users to decide if they want it readonly.

If we agree on how to handle creation attributes with this method I think I'll just overhaul the whole documentation on TypeScript (see https://github.com/sequelize/sequelize/pull/13909#issuecomment-1005820419)

ephys replied:
Done :)

ephys commented:
I've added a solution for creation attributes, but I went with a different direction than the branded type as it was causing issues.

Instead I went with a second option for `AttributesOf`: `optional`. This way it can be used for `TCreatrionAttributes` too:

```typescript
class User extends Model<
  AttributesOf<User, { omit: 'projects' }>,
  AttributesOf<User, { omit: 'projects', optional: 'id' }>
> {
  declare id: number;
  declare name: string;
  declare projects?: Project[];
}
```

Of course this double definition is due to backward compatibility requirements in v6.

In v7, I'll make it look like this instead:

```typescript
class User extends Model<User, { omit: 'projects', creationOptional: 'id' }> {
  declare id: number;
  declare name: string;
  declare projects?: Project[];
}
```

Unless we find a solution for branded types

ephys commented:
Oh no, I found a way to make a branded type that works.

I'll revert some things.

<img width="659" alt="image" src="https://user-images.githubusercontent.com/1280915/148560222-c9fb4ef3-86ea-476a-b150-fa1f3678d76d.png">

ephys commented:
I think this PR is almost ready:

- Added two utility types: `AttributesOf` & `CreationAttributesOf`
- Added two Branding types: `NonAttribute` & `CreationOptional`
- Overhauled the TypeScript documentation
- Had to mark all CreationAttributes that accept `undefined` as `optional` too, but I think that's a good change. :)
- Added lots of tests

Still needs to be done:

- Check that every instance of `_creationAttributes` is handled.
- Maybe replace `MakeUndefinedOptional<TModel['_creationAttributes']>` with a `GetCreationAttributes<TModel>` type to reduce duplication. (Note that it's different from `CreationAttributesOf`)
- Implement documentation updates from review

Going to put it back as draft until then.

Part of the Branding Type is a bit hacky. I've added lots of tests to make sure it doesn't negatively impact users. If anyone thinks of a scenario that would break don't hesitate to tell me.

ephys commented on docs/manual/other-topics/typescript.md:
```diff
@@ -296,12 +293,228 @@ async function doStuffWithUserModel() {
+  declare addProject: HasManyAddAssociationMixin<Project, number>;
+  declare hasProject: HasManyHasAssociationMixin<Project, number>;
+  declare countProjects: HasManyCountAssociationsMixin;
+  declare createProject: HasManyCreateAssociationMixin<Project, 'ownerId'>;
```
Added :)
Never realised there were that many. All the more reasons in favor of something like https://github.com/sequelize/meetings/issues/14

ephys commented:
Ok now I think I'm fully done.

We just need to decide for TypeScript < 4.1.

I'm also still uncertain about using Branded Types. It feels fragile. But so far it works for all the different types I've included.

ephys commented:
Well that's annoying, esbuild fails on numeric separators & optional catch bindings. I'm not sure why it didn't complain about it during the eslint PR. I'll revert these changes and we'll do them again once https://github.com/sequelize/sequelize/pull/13914 is finalized

WikiRik commented:
esbuild or esdoc?

ephys commented:
esdoc* :)

ephys commented:
I think I figured what the issue was. It's not related to the new linting rules. It just did not like seeing HTML comments in a markdown file.

ephys commented:
I think this is good to go now, just need a review 🙂

WikiRik commented on docs/manual/other-topics/typescript.md:
```diff
@@ -16,131 +17,177 @@ In order to avoid installation bloat for non TS users, you must install the foll
 
-// Some attributes are optional in `User.build` and `User.create` calls
-interface UserCreationAttributes extends Optional<UserAttributes, "id"> {}
+This solution is verbose. Sequelize v6 provides new utility types that will drastically reduce the amount
```
Maybe include which version of v6 adds this?

ephys replied:
I don't know which version of v6 it is yet but we can do this just before a merge :)

WikiRik commented on types/lib/associations/belongs-to-many.d.ts:
```diff
@@ -1,6 +1,6 @@
 import {
   BulkCreateOptions,
-  CreateOptions,
+  CreateOptions, CreationAttributes,
```
Maybe put it on a new line?

WikiRik commented on types/lib/associations/has-many.d.ts:
```diff
@@ -1,14 +1,15 @@
 import { DataType } from '../data-types';
 import {
-  CreateOptions,
+  CreateOptions, CreationAttributes,
```
And put this on a new line as well?

ephys replied:
I wish the linter would handle that, maybe at some point

WikiRik commented on types/lib/hooks.d.ts:
```diff
@@ -97,11 +97,17 @@ export class Hooks<
    * A similar dummy variable that doesn't exist on the real object. Do not
    * try to access this in real code.
+   *
+   * @deprecated This property will become a Symbol in v7 to prevent collisions.
```
Maybe add an expiring TODO for this?

WikiRik replied:
Or do you think marking it as deprecated is enough for us to remember to remove it when we get closer to the v7 betas?

ephys replied:
Done, added a TODO (although it won't be triggered because /types is ignored by eslint)

WikiRik commented on types/lib/sequelize.d.ts:
```diff
@@ -19,7 +19,7 @@ import {
   ModelCtor,
   Hookable,
-  ModelType,
+  ModelType, CreationAttributes, Attributes,
```
Maybe put these in three lines?

WikiRik reviewed (commented):
Looks good. Few small remarks. I want to wait for the approval of @AllAwesome497 before we merge

AllAwesome497 reviewed (approved):
looks fine to me

ephys commented on types/lib/hooks.d.ts:
```diff
@@ -97,13 +97,19 @@ export class Hooks<
+   * Use Attributes<Model> instead of this property to be forward-compatible.
    */
-  _attributes: TModelAttributes;
+  _attributes: TModelAttributes; // TODO [>6]: make this a non-exported symbol (same as the one in model.d.ts)
```
I can't yet, it would break user code, but I can in v7

WikiRik replied:
Maybe add that to the TODO as well?

ephys replied:
Now that I think of it, if it's `private`, `Attribute<M>` will not be able to access it. It's going to have to be a public internal Symbol I think

ephys commented:
@WikiRik small documentation update: I've added the v6 expected release version, fixed the documentation using the old utility types, and documented `Attributes<Model>`, `CreationAttributes<Model>`, and `ModelStatic<Model>`
